src/Microsoft.PowerShell.Commands.Utility/commands/utility/WebCmdlet/Common/WebRequestPSCmdlet.Common.cs

```diff
@@ -1265,106 +1280,133 @@ protected override void ProcessRecord()
-                                    request.Method,
-                                    request.RequestUri,
-                                    requestContentLength);
-                                WriteVerbose(reqVerboseMsg);
-
-                                HttpResponseMessage response = GetResponse(client, request, stripAuthorization);
-
-                                string contentType = ContentHelper.GetContentType(response);
-                                string respVerboseMsg = string.Format(CultureInfo.CurrentCulture,
-                                    WebCmdletStrings.WebResponseVerboseMsg,
-                                    response.Content.Headers.ContentLength,
-                                    contentType);
-                                WriteVerbose(respVerboseMsg);
-
-                                if (!response.IsSuccessStatusCode)
+                                FillRequestStream(request);
+                                try
                                 {
-                                    string message = String.Format(CultureInfo.CurrentCulture, WebCmdletStrings.ResponseStatusCodeFailure,
-                                        (int)response.StatusCode, response.ReasonPhrase);
-                                    HttpResponseException httpEx = new HttpResponseException(message, response);
-                                    ErrorRecord er = new ErrorRecord(httpEx, "WebCmdletWebResponseException", ErrorCategory.InvalidOperation, request);
-                                    string detailMsg = "";
-                                    StreamReader reader = null;
-                                    try
+                                    long requestContentLength = 0;
+                                    if (request.Content != null)
+                                        requestContentLength = request.Content.Headers.ContentLength.Value;
+
+                                    string reqVerboseMsg = String.Format(CultureInfo.CurrentCulture,
+                                        WebCmdletStrings.WebMethodInvocationVerboseMsg,
+                                        request.Method,
+                                        request.RequestUri,
+                                        requestContentLength);
+                                    WriteVerbose(reqVerboseMsg);
+                                    
+                                    response = GetResponse(client, request, stripAuthorization);
+                                    
+                                    bool causeFailureForTesting = (System.Management.Automation.Internal.InternalTestHooks.FailWebRequestCount > 0) && retriesLeft > 0;
+
+                                    isSuccessful = response.IsSuccessStatusCode && !causeFailureForTesting;                                
+
+                                    if (!isSuccessful && retriesLeft > 0)
```

Comment:
This current logic leads to an interestingly weird combination. If you set `-MaximumRedirection 0 -RetryCount 5 -RetryIntervalSec 1` and the response code is `301`, this will still retry the same link 5 times because `HttpResponseMessage.IsSuccessStatusCode` is only true for response codes `200`-`299` inclusive... not sure how to rectify that.

Reply:
also... since the logic for reducing the current redirection count is handled in `GetResponse()`, we may need to reset `WebSession.MaximumRedirection` to the `MaximumRedirection` property value before continuing.

Reply:
Best that I can find, the HTTP spec doesn't say enough about rate limiting to make it something we can depend on reliably.  However, I think we can say that retry only occurs on a 429 or 403 and document it as such unless there's data showing that web services are using other error codes.

Reply:
@SteveL-MSFT I can see other usecases for this beyond just rate limiting. I could see using this to test 404 errors until a new page comes online after a publish/push and it $x attempts fail alert that the push/publish failed. 

Perhaps it we limit it to 400-499 inclusive it would be best. In my experience with APIs I have seen more than 403 and 429 used for busy endpoints (including 404... Reddit... ugh) or rate limiting, but it's always a 400 level error

Reply:
@markekraus that's a fair point.  I think 4xx would be fine.

Reply:
After further consideration, this should work for the following status codes: 304. 400-599 inclusive

304 could be used for monitoring the change of a resource and is not a redirect status even though it in the 300 range. Retries on 5XX errors makes perfect sense to overcome temporary service glitches and cloudflare shenanigans.